x/warp/config.go

```diff
@@ -0,0 +1,213 @@
+	errInvalidAddressedPayload = errors.New("cannot unpack addressed payload")
+	errCannotGetNumSigners     = errors.New("cannot fetch num signers from warp message")
+)
+
+// Config implements the precompileconfig.Config interface and
+// adds specific configuration for Warp.
+type Config struct {
+	precompileconfig.Upgrade
+	QuorumNumerator uint64 `json:"quorumNumerator"`
+}
+
+// NewConfig returns a config for a network upgrade at [blockTimestamp] that enables
+// Warp with the given quorum numerator.
+func NewConfig(blockTimestamp *big.Int, quorumNumerator uint64) *Config {
+	return &Config{
+		Upgrade:         precompileconfig.Upgrade{BlockTimestamp: blockTimestamp},
+		QuorumNumerator: quorumNumerator,
+	}
+}
+
+// NewDefaultConfig returns a config for a network upgrade at [blockTimestamp] that enables
+// Warp with the default quorum numerator (0 denotes using the default).
+func NewDefaultConfig(blockTimestamp *big.Int) *Config {
+	return NewConfig(blockTimestamp, 0)
+}
+
+// NewDisableConfig returns config for a network upgrade at [blockTimestamp]
+// that disables Warp.
+func NewDisableConfig(blockTimestamp *big.Int) *Config {
+	return &Config{
+		Upgrade: precompileconfig.Upgrade{
+			BlockTimestamp: blockTimestamp,
+			Disable:        true,
+		},
+	}
+}
+
+// Key returns the key for the Warp precompileconfig.
+// This should be the same key as used in the precompile module.
+func (*Config) Key() string { return ConfigKey }
+
+// Verify tries to verify Config and returns an error accordingly.
+func (c *Config) Verify() error {
+	if c.QuorumNumerator > params.WarpQuorumDenominator {
+		return fmt.Errorf("cannot specify quorum numerator (%d) > quorum denominator (%d)", c.QuorumNumerator, params.WarpQuorumDenominator)
+	}
+	// If a non-default quorum numerator is specified and it is less than the minimum, return an error
+	if c.QuorumNumerator != 0 && c.QuorumNumerator < params.WarpQuorumNumeratorMinimum {
+		return fmt.Errorf("cannot specify quorum numerator (%d) < min quorum numerator (%d)", c.QuorumNumerator, params.WarpQuorumNumeratorMinimum)
+	}
+	return nil
+}
+
+// Equal returns true if [s] is a [*Config] and it has been configured identical to [c].
+func (c *Config) Equal(s precompileconfig.Config) bool {
+	// typecast before comparison
+	other, ok := (s).(*Config)
+	if !ok {
+		return false
+	}
+	equals := c.Upgrade.Equal(&other.Upgrade)
+	return equals && c.QuorumNumerator == other.QuorumNumerator
+}
+
+func (c *Config) Accept(acceptCtx *precompileconfig.AcceptContext, txHash common.Hash, logIndex int, topics []common.Hash, logData []byte) error {
+	unsignedMessage, err := warp.ParseUnsignedMessage(logData)
+	if err != nil {
+		return fmt.Errorf("failed to parse warp log data into unsigned message (TxHash: %s, LogIndex: %d): %w", txHash, logIndex, err)
+	}
+	log.Info("Accepted warp unsigned message", "txHash", txHash, "logIndex", logIndex, "logData", common.Bytes2Hex(logData))
```

Comment:
Isn't this signed and verified by `VerifyPredicate` at this point?

Reply:
No. When you send a message it produces a log. When you verify a message, it verifies a predicate and makes it read-able throughout the transaction's execution.